addons/crm/models/crm_lead.py

```diff
@@ -1241,6 +1259,65 @@ def _handle_salesmen_assignment(self, user_ids=False, team_id=False):
     # CLASSIFICATION TOOLS
     # --------------------------------------------------
 
+    def _get_potential_lead_duplicates_domain(self, lead, include_lost=False):
+        """ Returns a domain that can be used to retrieve the potential lead duplicates.
+            see: `_get_potential_lead_duplicates`
+            :param lead: Reference lead
+            :param boolean include_lost: If true, the leads considered as lost will be included in the results
+        """
+
+        domain = []
+        for normalized_email in filter(None, [tools.email_normalize(email) for email in tools.email_split(lead.email_from)]):
+            if '@' not in normalized_email:
+                continue
+            _, _, email_domain = normalized_email.rpartition('@')
+            if email_domain in iap_tools._MAIL_DOMAIN_BLACKLIST:
+                domain.append(('email_normalized', '=', normalized_email))
+            else:
+                domain.append(('email_normalized', '=ilike', '%@' + email_domain))
+
+        if lead.partner_id:
+            domain.append(('partner_id', '=', lead.partner_id.id))
+
+        if not domain:
+            return False
+
+        domain = expression.OR([[dom] for dom in domain])
+
+        if not include_lost:
+            domain = expression.AND([
+                domain,
+                expression.AND([
+                    [('active', '=', True)],
+                    expression.OR([
+                        [('type', '=', 'lead')],
+                        expression.AND([
+                            [('type', '=', 'opportunity')],
+                            expression.OR([
+                                [('probability', '=', False)],
+                                [('probability', '>', 0)]
+                            ])
+                        ])
+                    ])
+                ])
+            ])
+
+        return domain
+
+    def _get_potential_lead_duplicates(self, lead, include_lost=False):
+        """ Returns the leads that could be considered as "duplicate".
+        A lead will be considered as duplicate if:
+            - The lead is linked to the given partner
+            - The lead has a similar email address
+        If the provided email address has a common domain name (e.g: google.com, yahoo.com, ...),
+        the function will retrieve the leads having the same email address.
+        Otherwise, the function will retrieve the leads having the same email domain.
+        :param lead: Reference lead
+        :param boolean include_lost: If true, the leads considered as lost will be included in the results
+        """
+        domain = self._get_potential_lead_duplicates_domain(lead, include_lost)
+        return lead if domain is False else self.with_context(active_test=False).search(domain)
```

Comment:
If include_lost is False, do you need active_test = False ? As archived leads are considered as lost. Maybe for opportunities but is it necessary ?